### ISSUE TITLE
Montezuma's revenge

### ISSUE THREAD

**Issue by hlsafin** [yfletberliac/adversarially-guided-actor-critic#4]
I am testing this on Montezuma's revenge, and it doesn't seem to be doing too well, has this been tested? Are there any parameters I should test out for this env?

**Comment by yfletberliac** [comment]
Hi Safin,

We have not used AGAC for any of the Atari tasks, no.
I would recommend first making sure that the hyperparameters listed [here](https://github.com/DLR-RM/rl-baselines3-zoo/blob/643fdee2ce300e945e299a77508d47613e95e43e/hyperparams/ppo.yml#L1) for Atari are used, as the base algorithm that AGAC uses is PPO. And then tuning the AGAC-specific hyperparameters.

Please let us know your findings!

Yannis

**Comment by hlsafin** [comment]
```
env = make_doom_env("MontezumaRevengeNoFrameskip-v4", 64, seed, monitor_dir=log_dir)
model = AGAC('CnnPolicy', env, verbose=1, seed=seed, vf_coef=0.5, tensorboard_log=log_dir,
                n_steps=128, nminibatches=8, agac_c=linear_schedule(0.01), beta_adv=0.5,
                learning_rate=linear_schedule(2.5e-4), ent_coef=0.01, episodic_count=False)
model.learn(total_timesteps=1000000000, tb_log_name="tb/AGAC")
```

Currently, I have it like this.

**Comment by hlsafin** [comment]
![image](https://user-images.githubusercontent.com/36808133/175371521-44edf4e2-f8a7-4afc-9f24-5c571ac1892d.png)

This is Solaris so far, I'll keep you updated on this.

**Comment by yfletberliac** [comment]
Sounds good. How does the `agac_advantage` and `value_function_loss` look like?

Yannis

**Comment by hlsafin** [comment]
![image](https://user-images.githubusercontent.com/36808133/175801203-50cb0daf-116c-4543-bc7c-ba8677260886.png)
![image](https://user-images.githubusercontent.com/36808133/175801212-32578226-1159-449c-9e5a-67d690cbaf5c.png)
Yeah, It didn't look like it was improving after 25M steps, so I stopped it. What do you think? Too early?

**Comment by yfletberliac** [comment]
Thanks for the update!

Have you also changed the `cliprange` to 0.1 (or linear_schedule(0.1))?
Since linear scheduling is influenced by `total_timesteps`, I would try to use a smaller `total_timesteps` than what you currently have (from https://github.com/yfletberliac/adversarially-guided-actor-critic/issues/4#issuecomment-1164702418 it appears to be 1000000000), e.g. 1e7 or 5e7.
Maybe the only unexpected element from the graphs is the value function loss which appears to be quite high compared to the policy gradient loss...

Yannis

**Comment by hlsafin** [comment]
model = AGAC('CnnPolicy', env, verbose=1, seed=seed, vf_coef=0.5, tensorboard_log=log_dir,
                n_steps=128, nminibatches=8, agac_c=linear_schedule(0.01), beta_adv=0.5,cliprange=linear_schedule(0.1),
                learning_rate=linear_schedule(2.5e-4), ent_coef=0.01, episodic_count=False)
model.learn(total_timesteps=int(5e7), tb_log_name="tb/AGAC")

**Comment by hlsafin** [comment]
![image](https://user-images.githubusercontent.com/36808133/176230043-b9ee8b38-a233-4bba-a0c1-c9ce3839b2ae.png)
![image](https://user-images.githubusercontent.com/36808133/176230102-529aa0e5-9020-43ff-bc48-12a4976b7fd0.png)
Here is the result, it somewhat did worse then previous parameters.

**Comment by yfletberliac** [comment]
Ok thanks. If you have the resources could you try setting `agac_c` higher?

**Comment by hlsafin** [comment]
what would be a good agac_c value?

**Comment by yfletberliac** [comment]
Perhaps an order of magnitude higher so `agac_c=linear_schedule(0.1)`.

**Comment by hlsafin** [comment]
![image](https://user-images.githubusercontent.com/36808133/177050565-8e73577b-1de9-437a-ac48-b3576e35fc8f.png)

![image](https://user-images.githubusercontent.com/36808133/177050579-83d06ed1-27c9-4b1c-9e6c-14f995fb57f8.png)

**Comment by hlsafin** [comment]
It performed pretty much on par with other experiments.

**Comment by yfletberliac** [comment]
The value function loss still appears to be extremely high compared to the policy gradient loss.
One way to tackle this would be to do 5-10 critic/value function updates each time one actor/policy update is made, with the same batch of data.
The change would need to be made inside `_train_step`, [here](https://github.com/yfletberliac/adversarially-guided-actor-critic/blob/7ed97711e4ff70e33f07da01320d6f2b080acae2/agac/agac.py#L301) and [here](https://github.com/yfletberliac/adversarially-guided-actor-critic/blob/7ed97711e4ff70e33f07da01320d6f2b080acae2/agac/agac.py#L311) to make the `self.vf_loss` update 5-10 times while everything else gets updated once. This should put the `loss/value_function_loss` down.

**Comment by hlsafin** [comment]
To be frank, I'm not too familiar with the Tensorflow gradient update, and should this mean that the actor and critic are decoupled?

**Comment by yfletberliac** [comment]
Hi Safin,

Could you elaborate on what you mean by decoupled? If you mean independent, then yes the actor and the critic are already two independent neural nets.

You would need to put something like this in place of lines 301 to 320 (in the code snippets below I've just added the two first lines):
```
for _ in range(4):
            value_loss = self.sess.run([self.vf_loss], td_map)
if writer is not None:
            # Run loss backprop with summary, but once every 10 runs save the metadata (memory, compute time, ...)
            if self.full_tensorboard_log and (1 + update) % 10 == 0:
                run_options = tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE)
                run_metadata = tf.RunMetadata()
                summary, policy_loss, value_loss, policy_entropy, approxkl, clipfrac, _ = self.sess.run(
                    [self.summary, self.pg_loss, self.vf_loss, self.entropy, self.approxkl, self.clipfrac,
                     self._train],
                    td_map, options=run_options, run_metadata=run_metadata)
                writer.add_run_metadata(run_metadata, 'step%d' % (update * update_fac))
            else:
                summary, policy_loss, value_loss, policy_entropy, approxkl, clipfrac, _ = self.sess.run(
                    [self.summary, self.pg_loss, self.vf_loss, self.entropy, self.approxkl, self.clipfrac,
                     self._train],
                    td_map)
            writer.add_summary(summary, (update * update_fac))
        else:
            policy_loss, value_loss, policy_entropy, approxkl, clipfrac, _ = self.sess.run(
                [self.pg_loss, self.vf_loss, self.entropy, self.approxkl, self.clipfrac,
                 self._train], td_map)
```
This will update the critic/value function network 4 times before updating everything (hence a total of 5 critic updates while the actor is updated only once).

**Comment by hlsafin** [comment]
![image](https://user-images.githubusercontent.com/36808133/177694077-a45cd728-21e0-4578-b12e-ed998fd4a804.png)
![image](https://user-images.githubusercontent.com/36808133/177694098-f94f0639-7fd7-46bb-85f2-e776513f03b3.png)
Here is the result so far for 6 million timestep, it seems like the value loss is still exploding, no?

**Comment by hlsafin** [comment]
But I might be a bit too early, so I will let it train until 20-30 mil and see how it behaves.

**Comment by yfletberliac** [comment]
It seems to not be helping indeed.
Alternatively, there seem to be another set of hyperparameter in this [paper](https://arxiv.org/abs/2104.06159) using PPO (and an improved version of it). I have a screenshot of Table 5 below:

<img width="442" alt="image" src="https://user-images.githubusercontent.com/14262199/177702578-3123de05-a39b-4b1e-a00d-c58bc3a07a0e.png">

Hyperparameters such as learning rate are different (still annealed but starting at `3e-4`), batch size, or discount factor. An important change is also that they put themselves in the off-policy setting by allowing the use of a replay buffer - which might be necessary (even with an AGAC adversary) with a task such as Montezuma with a long and forgettable sequence of events...

**Comment by hlsafin** [comment]
![image](https://user-images.githubusercontent.com/36808133/177862703-9ba865ee-eb05-40cf-919d-6d553da0a994.png)
![image](https://user-images.githubusercontent.com/36808133/177862719-59d66a7f-31ef-4c9f-a593-5abbd7c75897.png)
Here is at 30 mil.

**Comment by hlsafin** [comment]
This is Solaris, which I think its a bit easier then montezuma;s revenge.
